src/_pytest/recwarn.py

```diff
@@ -206,12 +206,23 @@ def __len__(self) -> int:
         return len(self._list)
 
     def pop(self, cls: Type[Warning] = Warning) -> "warnings.WarningMessage":
-        """Pop the first recorded warning, raise exception if not exists."""
+        """Pop the first recorded warning (or subclass of warning), raise exception if not exists."""
+        matches = []
         for i, w in enumerate(self._list):
-            if issubclass(w.category, cls):
+            if w.category == cls:
                 return self._list.pop(i)
-        __tracebackhide__ = True
-        raise AssertionError(f"{cls!r} not found in warning list")
+            if issubclass(w.category, cls):
+                matches.append((i, w))
+        if not matches:
+            __tracebackhide__ = True
+            raise AssertionError(f"{cls!r} not found in warning list")
+        (idx, best), *rest = matches
+        for i, w in rest:
+            if issubclass(w.category, best.category) and not issubclass(
+                best.category, w.category
+            ):
+                idx, best = i, w
+        return self._list.pop(idx)
```

Comment:
I think we could do this in a single pass, better than my earlier sketch:

```python
best_idx = None
for i, w in enumerate(self._list):
    if w.category == cls:
        return self._list.pop(i)  # exact match, stop looking
    if issubclass(w.category, cls) and (
        best_idx is None or not issubclass(w.category, self._list[best_idx].category)
    ):
        best_idx = i
if best_idx is not None:
    return self._list.pop(best_idx)
__tracebackhide__ = True
raise AssertionError(f"{cls!r} not found in warning list")
```

Reply:
Interesting mypy is screaming that right side of or will never be evaluated
```python
or not issubclass(w.category, self._list[best_idx].category)
```
all tests works also without it

Reply:
the mypy error is exactly:
```bash
src/_pytest/recwarn.py:222: error: Right operand of "or" is never evaluated  [unreachable]
```

Reply:
Mypy must be assuming that best_idx is always None, and is definitely wrong about this.  If we don't have a failing test we should fix that.

Reply:
I added it, so I will try to figure out how to overcome mypy

Reply:
I put there `# type: ignore[unreachable]` it is a problem of type narrowing, that has a solution to add there elseif or assert that it is really not a None